In [1]:
# In the Data cleaning part pandas library is most preferable
import pandas as pd
import numpy as np

In [2]:
#Impoting the file which we are done in scraping part using read_excel method
df = pd.read_excel('df.xlsx')

In [3]:
df.head()

Unnamed: 0  Bank_Name      Bank_Type       Location  \
0           0  HDFC Bank  Personal Loan        , salem   
1           1        SBI  Personal Loan    , proddatur   
2           2  HDFC Bank  Personal Loan    , proddatur   
3           3  Axis Bank  Personal Loan  , pudukkottai   
4           4  HDFC Bank  Personal Loan  , pudukkottai   

                          Bank_Caption    User_Name  Bank_Rating  \
0             "Need a quick response "     PANDIYAN          4.0   
1                       "Best service"        KUMAR          5.0   
2                    "Average service"        PAVAN          4.0   
3  "Need to improve customer services"  DEVASAHAYAM          1.0   
4                       "Good service"      SELVIN           5.0   

  Rating_Comment                                       User_Comment  
0          Great  \n                                I was applie...  
1          Blown  \n                                I have taken...  
2          Great  \n                                I was applie...  
3            Bad  \n                                I have take ...  
4          Blown  \n                                I have taken...

In [4]:
df.tail()

Unnamed: 0   Bank_Name       Bank_Type     Location  \
11975        3995         SBI  Saving Account  , bangalore   
11976        3996  ICICI Bank  Saving Account  , bangalore   
11977        3997      review      Other Loan  , bangalore   
11978        3998   New India             NaN  , bangalore   
11979        3999         Max  Life Insurance  , bangalore   

                 Bank_Caption User_Name  Bank_Rating Rating_Comment  \
11975         "Horrible bank"   ANONYMS          2.0            NaN   
11976          "Good Account"   ANONYMS          4.0          Great   
11977          "Good Account"   ANONYMS          4.0          Great   
11978  "Recommendable policy"   ANONYMS          4.0          Great   
11979           "Good Policy"   ANONYMS          4.0          Great   

                                            User_Comment  
11975  \n                                SBI is not a...  
11976  \n                                From 2014 on...  
11977  \n                                I am using t...  
11978  \n                                My employer ...  
11979  \n                                Taken my lif...

In [5]:
# when checking the data set it contain unnamed column and we drop that column with using drop() method
df.drop(columns = 'Unnamed: 0',inplace = True)
df.head()

Bank_Name      Bank_Type       Location  \
0  HDFC Bank  Personal Loan        , salem   
1        SBI  Personal Loan    , proddatur   
2  HDFC Bank  Personal Loan    , proddatur   
3  Axis Bank  Personal Loan  , pudukkottai   
4  HDFC Bank  Personal Loan  , pudukkottai   

                          Bank_Caption    User_Name  Bank_Rating  \
0             "Need a quick response "     PANDIYAN          4.0   
1                       "Best service"        KUMAR          5.0   
2                    "Average service"        PAVAN          4.0   
3  "Need to improve customer services"  DEVASAHAYAM          1.0   
4                       "Good service"      SELVIN           5.0   

  Rating_Comment                                       User_Comment  
0          Great  \n                                I was applie...  
1          Blown  \n                                I have taken...  
2          Great  \n                                I was applie...  
3            Bad  \n                                I have take ...  
4          Blown  \n                                I have taken...

#### In the Location column we have (), extra so we are removed that special character using strip method

In [6]:

df['Location']=df['Location'].str[1:]

In [7]:
df.head()

Bank_Name      Bank_Type      Location  \
0  HDFC Bank  Personal Loan         salem   
1        SBI  Personal Loan     proddatur   
2  HDFC Bank  Personal Loan     proddatur   
3  Axis Bank  Personal Loan   pudukkottai   
4  HDFC Bank  Personal Loan   pudukkottai   

                          Bank_Caption    User_Name  Bank_Rating  \
0             "Need a quick response "     PANDIYAN          4.0   
1                       "Best service"        KUMAR          5.0   
2                    "Average service"        PAVAN          4.0   
3  "Need to improve customer services"  DEVASAHAYAM          1.0   
4                       "Good service"      SELVIN           5.0   

  Rating_Comment                                       User_Comment  
0          Great  \n                                I was applie...  
1          Blown  \n                                I have taken...  
2          Great  \n                                I was applie...  
3            Bad  \n                                I have take ...  
4          Blown  \n                                I have taken...

#### In the User_Comment column to delete (\n) again we use strip() method

In [8]:
df['User_Comment']=df['User_Comment'].str[3:]

In [9]:
df.head()

Bank_Name      Bank_Type      Location  \
0  HDFC Bank  Personal Loan         salem   
1        SBI  Personal Loan     proddatur   
2  HDFC Bank  Personal Loan     proddatur   
3  Axis Bank  Personal Loan   pudukkottai   
4  HDFC Bank  Personal Loan   pudukkottai   

                          Bank_Caption    User_Name  Bank_Rating  \
0             "Need a quick response "     PANDIYAN          4.0   
1                       "Best service"        KUMAR          5.0   
2                    "Average service"        PAVAN          4.0   
3  "Need to improve customer services"  DEVASAHAYAM          1.0   
4                       "Good service"      SELVIN           5.0   

  Rating_Comment                                       User_Comment  
0          Great                                I was applied HD...  
1          Blown                                I have taken a p...  
2          Great                                I was applied HD...  
3            Bad                                I have take a pe...  
4          Blown                                I have taken a t...

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11980 entries, 0 to 11979
Data columns (total 8 columns):
Bank_Name         11980 non-null object
Bank_Type         11589 non-null object
Location          11978 non-null object
Bank_Caption      11980 non-null object
User_Name         11965 non-null object
Bank_Rating       11980 non-null float64
Rating_Comment    10938 non-null object
User_Comment      11980 non-null object
dtypes: float64(1), object(7)
memory usage: 748.8+ KB


In [11]:
df.columns

Index(['Bank_Name', 'Bank_Type', 'Location', 'Bank_Caption', 'User_Name',
       'Bank_Rating', 'Rating_Comment', 'User_Comment'],
      dtype='object')

### Observations:

- The shape of the dataset 11980 rows and 8 columns.
- Presenting the data set having the column names as 
        - Bank_Name: Name of the banks in india
        - Bank_Type: Type of the bank account
        - Location: City of the Banks
        - Bank_Caption: Motto line of every bank
        - User_Name: name of the person who are given a rating for the bank
        - Bank_Rating: Rating of the bank by the user_name
        - Rating_Comment: Commenting in a single line by the username
        - User_Comment: it contains the experience of the User_Name
- At last we observe that all the columns are having data type as object except Bank_Rating column, it is a float data type

In [12]:
# the data set contains some duplicate oservations so we removed those duplicates with using drop_duplicates()
df.drop_duplicates(inplace = True)
df.shape

(11107, 8)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11107 entries, 0 to 11979
Data columns (total 8 columns):
Bank_Name         11107 non-null object
Bank_Type         10743 non-null object
Location          11105 non-null object
Bank_Caption      11107 non-null object
User_Name         11094 non-null object
Bank_Rating       11107 non-null float64
Rating_Comment    10135 non-null object
User_Comment      11107 non-null object
dtypes: float64(1), object(7)
memory usage: 781.0+ KB


In [14]:
df.isnull().sum()

Bank_Name           0
Bank_Type         364
Location            2
Bank_Caption        0
User_Name          13
Bank_Rating         0
Rating_Comment    972
User_Comment        0
dtype: int64

### After checking the null values the data set having Null values in the Bank_Type column as 364 and Rating_comment having 972 missing values.

**For that we use row operation by using under the condition**

In [19]:
df = df[df.isnull().sum(axis=1)<1]

In [20]:
df[df.isnull().sum(axis=1)<2]

Bank_Name       Bank_Type      Location  \
0                   HDFC Bank   Personal Loan         salem   
1                         SBI   Personal Loan     proddatur   
2                   HDFC Bank   Personal Loan     proddatur   
3                   Axis Bank   Personal Loan   pudukkottai   
4                   HDFC Bank   Personal Loan   pudukkottai   
5                   HDFC Bank   Personal Loan   pudukkottai   
6                   HDFC Bank   Personal Loan     bangalore   
7                Hero fincorp     Two Wheeler    dharmapuri   
8                   HDFC Bank     Credit Card        mysore   
9                   HDFC Bank   Personal Loan        mysore   
10                  HDFC Bank   Personal Loan     bangalore   
11                        SBI  Saving Account     hyderabad   
12                  HDFC Bank     Credit Card     hyderabad   
13                  HDFC Bank   Personal Loan     hyderabad   
14                     review  Saving Account     hyderabad   
15                     review      Other Loan     hyderabad   
16                  HDFC Bank   Personal Loan     hyderabad   
17     HDB Financial Services   Personal Loan         surat   
18                  HDFC Bank   Personal Loan         surat   
19                  Axis Bank  Saving Account           goa   
37                  Axis Bank   Personal Loan           goa   
38                  HDFC Bank   Personal Loan           goa   
39                     review      Other Loan       auraiya   
41                  HDFC Bank     Credit Card        meerut   
42                     review      Other Loan        meerut   
43                     review     Credit Card        meerut   
44                     review      Other Loan        nagpur   
45              Bajaj Finserv   Personal Loan        nagpur   
46                  SBI Cards     Credit Card        nagpur   
47                   RBL Bank     Credit Card     hyderabad   
...                       ...             ...           ...   
11948             Iffco Tokio     Two Wheeler    coimbatore   
11949                     SBI  Saving Account     hyderabad   
11950               HDFC Bank  Saving Account     hyderabad   
11951                     SBI  Saving Account    coimbatore   
11952               HDFC Bank   Personal Loan     hyderabad   
11953               HDFC Bank     Credit Card     hyderabad   
11954                Citibank     Credit Card     hyderabad   
11955                  review  Saving Account    coimbatore   
11956                  review     Two Wheeler    coimbatore   
11957                  review      Other Loan    coimbatore   
11958                   Kotak     Credit Card    coimbatore   
11959               HDFC Bank     Credit Card    coimbatore   
11960               New India     Two Wheeler     hyderabad   
11961               HDFC Bank  Saving Account     hyderabad   
11962               HDFC Bank     Credit Card     hyderabad   
11963               SBI Cards     Credit Card     hyderabad   
11964               SBI Cards     Credit Card     hyderabad   
11965                Citibank     Credit Card     hyderabad   
11966           IndusInd Bank     Credit Card       lucknow   
11967                Citibank     Credit Card     hyderabad   
11968           IndusInd Bank     Credit Card           new   
11969                Citibank  Saving Account     bangalore   
11970                     LIC  Life Insurance     bangalore   
11971               HDFC Bank     Credit Card     bangalore   
11972                Citibank     Credit Card     bangalore   
11973               HSBC Bank     Credit Card     bangalore   
11974                  review      Other Loan     bangalore   
11976              ICICI Bank  Saving Account     bangalore   
11977                  review      Other Loan     bangalore   
11979                     Max  Life Insurance     bangalore   

                                  Bank_Caption              User_Name  \
0                     "Need a quick resp

In [21]:
len(df[df.isnull().sum(axis=1)<3])

9790

In [22]:
df.shape

(9790, 8)

In [23]:
df["Rating_Comment"].value_counts()

Blown           4571
Great           3172
Satisfactory    1397
Bad              399
Pretty           251
Name: Rating_Comment, dtype: int64

#### By the row operation all the null_values are deleted. Now the data set containing 9790 rows and 8 columns.

In [24]:
df.isnull().sum()

Bank_Name         0
Bank_Type         0
Location          0
Bank_Caption      0
User_Name         0
Bank_Rating       0
Rating_Comment    0
User_Comment      0
dtype: int64

In [25]:
df.shape

(9790, 8)

In [26]:
df.head()

Bank_Name      Bank_Type      Location  \
0  HDFC Bank  Personal Loan         salem   
1        SBI  Personal Loan     proddatur   
2  HDFC Bank  Personal Loan     proddatur   
3  Axis Bank  Personal Loan   pudukkottai   
4  HDFC Bank  Personal Loan   pudukkottai   

                          Bank_Caption    User_Name  Bank_Rating  \
0             "Need a quick response "     PANDIYAN          4.0   
1                       "Best service"        KUMAR          5.0   
2                    "Average service"        PAVAN          4.0   
3  "Need to improve customer services"  DEVASAHAYAM          1.0   
4                       "Good service"      SELVIN           5.0   

  Rating_Comment                                       User_Comment  
0          Great                                I was applied HD...  
1          Blown                                I have taken a p...  
2          Great                                I was applied HD...  
3            Bad                                I have take a pe...  
4          Blown                                I have taken a t...

In [27]:
df.to_csv('Clean_data_set.csv')


## Data_Understanding

In [28]:
df.keys()

Index(['Bank_Name', 'Bank_Type', 'Location', 'Bank_Caption', 'User_Name',
       'Bank_Rating', 'Rating_Comment', 'User_Comment'],
      dtype='object')

### Bank_Name column

In [29]:
df['Bank_Name'].unique()

array(['HDFC Bank', 'SBI', 'Axis Bank', 'Hero fincorp', 'review',
       'HDB Financial Services', 'Bajaj Finserv', 'SBI Cards', 'RBL Bank',
       'Kotak', 'IndusInd Bank', 'LIC', 'ICICI Bank', 'Citibank',
       'New India', 'andhra Bank', 'Punjab National Bank', 'YES Bank',
       'Canara Bank', 'ICICI Lombard', 'HSBC Bank', 'Sundaram Finance',
       'American Express', 'Fullerton', 'Tata AIA', 'IDFC Bank',
       'Birla Sun', 'Exide', 'Standard Chartered Bank',
       'Bajaj Allianz Life', 'Oriental Bank of Commerce', 'Shriram City',
       'Iffco Tokio', 'Tata capital', 'Incred', 'Toyota Finance',
       'Federal Bank', 'Cashe', 'ICICI', 'Indian Bank', 'Shubh Loans',
       'HDFC Ergo', 'United India', 'Cholamandalam Finance',
       'Capital First', 'Bajaj Allianz', 'Shriram', 'SBI Life',
       'Reliance General', 'ICICI Prudential', 'National Insurance',
       'IDBI', 'Karnataka Bank', 'Max', 'Bank of Baroda', 'Shriram Life',
       'Royal Sundaram', 'Oriental Insurance', 'L&

In [30]:
#pd.pivot_table(df,index = ["Bank_Type","Rating_Comment"])

In [31]:
df['Bank_Name'].nunique()

86

In [32]:
df['Bank_Name'].value_counts()

HDFC Bank                                2196
review                                   1566
IndusInd Bank                            1459
Citibank                                  643
ICICI Bank                                521
SBI Cards                                 489
Axis Bank                                 400
SBI                                       364
Kotak                                     307
Fullerton                                 267
LIC                                       212
YES Bank                                  203
RBL Bank                                  175
Incred                                    134
HSBC Bank                                 133
Canara Bank                                56
Tata capital                               51
Bajaj Finserv                              40
American Express                           37
Standard Chartered Bank                    37
ICICI Lombard                              34
Punjab National Bank              

#### In the Bank_Name column 86 unique observations are there, in that HDFC bank repeated most of the times and the count is 2196.
#### In the same column the second most repeated bank is review i.e, bank bazaar web provided name it has the count 1566.
#### The data type of this column is object 

### Bank_Type_Column

In [33]:
df['Bank_Type'].unique()

array(['Personal Loan', 'Two Wheeler', 'Credit Card', 'Saving Account',
       'Other Loan', 'Life Insurance', 'Car Loan', 'Car Insurance',
       'Consumer Durable', 'Current Account', 'Loan Against',
       'Travel Insurance'], dtype=object)

In [34]:
df['Bank_Type'].nunique()

12

In [35]:
df['Bank_Type'].value_counts()

Credit Card         4156
Personal Loan       2423
Saving Account      1453
Other Loan           868
Life Insurance       299
Two Wheeler          249
Consumer Durable     196
Car Loan              79
Car Insurance         52
Current Account       11
Loan Against           3
Travel Insurance       1
Name: Bank_Type, dtype: int64

- The `Bank_Type` column having 12 unique types
- The `Bank_Type` colimn having credit card type is most frequency the count is 4156
- the second highest type is Personal Loan, the count is 2423
- Travel Insurance having minimum frequency as 1

### Location_Column

In [36]:
df['Location'].unique()

array([' salem', ' proddatur', ' pudukkottai', ' bangalore',
       ' dharmapuri', ' mysore', ' hyderabad', ' surat', ' goa',
       ' auraiya', ' meerut', ' nagpur', ' jaipur', ' vijayawada',
       ' surendranagar', ' gurgaon', ' jodhpur', ' pune', ' dhule',
       ' gwalior', ' mumbai', ' new', ' rajkot', ' ludhiana',
       ' visakhapatnam', ' raigarh', ' rajahmundry', ' hospet',
       ' hindupur', ' tirupur', ' kochi', ' haveri', ' tumkur',
       ' bareilly', ' warangal', ' tiruchirapalli', ' bellary',
       ' coimbatore', ' jalna', ' chennai', ' secunderabad', ' beawar',
       ' ajmer', ' jhunjhunu', ' alwar', ' dindigul', ' noida', ' rohtak',
       ' kanchipuram', ' ahmedabad', ' thiruvananthapuram', ' lucknow',
       ' thane', ' tirupati', ' guntur', ' indore', ' ernakulam',
       ' bhubaneshwar', ' navi', ' kolkata', ' gandhinagar', ' vellore',
       ' nellore', ' faridabad', ' thanjavur', ' cuttack', ' madurai',
       ' hosur', ' ghaziabad', ' howrah', ' aligarh', ' 

In [37]:
df['Location'].nunique()

293

In [38]:
df['Location'].value_counts()

 bangalore         1705
 hyderabad         1439
 chennai           1062
 mumbai             607
 new                595
 coimbatore         360
 kolkata            344
 pune               281
 thane              183
 ahmedabad          178
 gurgaon            130
 jaipur             119
 noida              105
 tirupur             86
 visakhapatnam       83
 kochi               75
 erode               72
 vijayawada          64
 surat               63
 lucknow             61
 faridabad           60
 indore              59
 madurai             59
 ghaziabad           58
 vadodara            56
 mysore              55
 navi                53
 tiruchirapalli      48
 nellore             45
 guntur              38
                   ... 
 ariyalur             1
 nainital             1
 manawar              1
 sedam                1
 auraiya              1
 krishnanagar         1
 sirohi               1
 hoshiarpur           1
 kannur               1
 shimoga              1
 jalalabad      

- In the `Location` column bangalore city having the most count i.e, 1705
- The second highest column city is hyderabad, the count is 1439
- The `Location` column having 293 unique cities
- Data type of the `Location` column is `object`

### Bank_Rating_Column

In [39]:
df['Bank_Rating'].unique()

array([4. , 5. , 1. , 3. , 3.5, 1.5])

In [40]:
df['Bank_Rating'].nunique()

6

In [41]:
df['Bank_Rating'].value_counts()

5.0    4571
4.0    3172
3.0    1397
1.0     390
3.5     251
1.5       9
Name: Bank_Rating, dtype: int64

- `Bank_Rating` column having 6 unique values 
- In the `Bank_Rating` column 5.0 rating has repeated most number of times it has 4571 count
- Half of the data has 5.0 rating
- The seecond highest is 4.0 it has count 3172
- So from this column 80% of data is above 4.0 rating
- This column having the `float` data type 
- The least rating is 1.5 count is 9

### Rating_Comment_Column

In [38]:
df['Rating_Comment'].unique()

array(['Great', 'Blown', 'Bad', 'Satisfactory', 'Pretty'], dtype=object)

In [39]:
df['Rating_Comment'].nunique()

5

In [40]:
df['Rating_Comment'].value_counts()

Blown           4571
Great           3172
Satisfactory    1397
Bad              399
Pretty           251
Name: Rating_Comment, dtype: int64

- `Bank_Rating` column having 5 unique values
- The highest value is `Blown(Away)`, it has `4571` count
- second highest value is `Great`, it has `3172` counnt
- the min value is `Pretty`, it has `251` count
- the data type of this column is `object`

In [41]:
df['User_Comment'].nunique()

9780